usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -5111,12 +5113,177 @@ zfs_get_holds(zfs_handle_t *zhp, nvlist_t **nvl)
 }
 
 /*
- * Convert the zvol's volume size to an appropriate reservation.
+ * The theory of raidz space accounting
+ *
+ * With raidz{1,2,3} vdevs, the referenced property's value is scaled relative
+ * to the amount of space required to store a 128k (SPA_OLD_MAXBLOCKSIZE) block.
+ * This is referred to as "raidz deflation."
+ *
+ * No matter how many blocks are allocated for a newly written uncompressed 128k
+ * block, referenced will increase by 128k.  However, the pool's allocated and
+ * free properties will be adjusted by the amount space allocated for data,
+ * parity, and perhaps skip blocks.  vdev_raidz_asize() is used for calculating
+ * the total amount of space required for a single block.  If another block size
+ * is written, the change to the referenced property may be different than the
+ * amount of space actually alloctaed.
```

Comment:
Great comment!  I think this one paragraph could be rewritten to emphasize the points that are salient here.  Here's a start:

> The "referenced" property of RAIDZ vdevs is scaled such that a 128KB block will "reference" 128KB, even though it allocates more than that, to store the parity information (and perhaps skip sectors).  This concept of the "referenced" (and other DMU space accounting) being lower than the allocated space by a constant factor is called "raidz deflation."
>
> As mentioned above, the constant factor for raidz deflation assumes a 128KB block size.  However, zvols typically have a much smaller block size (default 8KB).  These smaller blocks may require proportionally much more parity information (and perhaps skip sectors). In this case, the change to the "referenced" property may be much more than the logical block size.

Reply:
Thanks, that looks great.